Ulims 395 materialized schemas

mc7 opened the pull request:
This PR is being split into smaller PRd. It has been retained for reference .

A minimal set of files for defining JSON schemas, that are managed within using jsonschema-tool (written by wikipedia) with readme for writers.

Changes

- Added files need to use jsonschema-tools
- Updated repo read me to include links to schema writers page and view of schema
- Added new read me for schema writers, containing 4 tutorials that build on each other.
- Added "empty" schema folders for copying by writers when creating new schema.
- Added five schema with a 1.0.0 completed release that are allowed to break when merged to main. All other schema must never break when merged to main.
- Added files to cause Github to do schema root check on push or pull request.

Things to check

- [ ] Confirm build install works: npm install
- [ ] Do a SRC (schema root check) using: npm run check
- [ ] Confirm the github action passes (it does the same thing as above)
- [ ] Try the applications and schema writers  views of a schema using the links near top of repo README.
- [ ] Complete tutorial 1, which creates a schema with no $ref
- [ ] Optionally, complete tutorial 2 (schema with $ref to tutorial 1 schema) and tutorial 3 (changing $id and $ref on the two schema over time)
- [ ] Read text of tutorial 4 which applies tutorial 3 to main using the schema added in this commit. Can be done after this PR is merged.

The following tickets are caused by this PR, where level 3 requires this PR to be merged.

- ULIMS-431 will give summary of how repo works and highlight the tutorials.
- ULIMS-429 is about someone new to repo creating a new schema following tutorials 1-3
- ULIMS-430 is about someone new to repo doing tutorial 4. 

Tutorials 1-3 should be completed by all schema writers. Tutorial 4 can be done any schema writer, but is aimed at people doing PR reviews and merging to main.

ianbush commented on check.py:
We have a ticket to remove the Python code from here, so we should avoid updating this code.

ianbush reviewed (changes requested):
Generally looks like the `jsonschema-tools` are doing the job we want, which is good to see.

There are a lot of examples in here, but I do not think we need this variety. Can we please simplify this down to a single example, perhaps just the collection in `pattern9/faster`, and just rename it to test schema. The top level directly for schemas I suggest we call `schemas`, then we can have the single example folder in `schemas/test`. We should not need the `shared` or `instrument` folders yet.

The yaml files - can we get rid of these? I realise that the jsonschema-tools package likes to use them, but I think it would be best we stick to JSON avoid confusion with yaml files.

The GitHub actions runs the check, can we also have it generate the materialised schemas? This could be done as a separate ticket to keep this PR smaller.

The readme.md files in the individual examples can probably be consolidated with the general README.md for now. The glossary especially introduces a lot of terms, but I think many of these are adding extra complications we do not need. Would it be useful to simplify this as something like:
* Schema - e.g. `current.json` file that can have references to other schemas
* Built Schema - e.g. `0.0.1.json` that has references to other schemas resolved, and will be used by ULIMS applications

I don't think we need to formalise the schema root/schema set/schema group terms for now, though perhaps we would want something like this in the future as we build up a more complete set of schemas.

mc7 commented:
Thanks for the useful comments Ian. I've pushed changes to branch and edited the PR comment taking them into account. In summary:

- Revised the repo readme. In particular removed the glossary and linked to new page. The application view of a schema was changed and a writers view of the same schema was added.
- Changed to a single page for schema writers instead of the information being scattered around on different pages and on PR comment. That information was put under four tutorials that build on each other. The first 3 can be done now whilst the 4th needs this PR to be done.
- Linked tutorials to start of "fixes" page to be populated by future ticket(s).
- Emphasised the json first method of defining schema. De-emphasised set and group folders for managing schema. Stripped back the "empty" schema folders read me to minimum with placeholders.
- All "pattern" schema and Python added were have been deleted. One of the patterns had an undocumented (now in tutorial 4) steps applied and doing after this PR will make it clearer.
- Added schema group reposhow that has three purposes containing five schema with 1 release, needed for tutorial 4. The group readme explains why each exists and highlights these are they only schema allowed to break when a merge to main occurs.

ianbush reviewed (changes requested):
Can we do the following please:
- [ ] Rename `sroot` to `schemas`
- [ ] Remove the `.yaml` files (double check if we can do this)
- [ ] Leave just one example under `schemas/example`, with perhaps a minor version (0.0.2) and a major version change (1.0.0)
- [ ] Remove the `instruments` folder
- [ ] Remove the `practice folder`
- [ ] Move all the documentation into docs - I have not had a chance to look through this yet, but I think we need to look at significantly simplifying this

mc7 commented:
>
>     * [ ]  Remove the `.yaml` files (double check if we can do this)
> 
Double checked the blank yaml file needs to exist else SRC: (npm run check) will fails
![image](https://github.com/user-attachments/assets/4ab603a8-2435-4467-81d8-c4080ee2642a)

ianbush commented on docs/glossary.md:
I am not sure we need the glossary at the moment. In order to get the main bits of this branch merged can we please move this to a separate docs branch and consider it further later? Then the info in the `readme.md` can just cover the main points. I've made a similar comment on the tutorials, they can go on the same new docs branch.

mc7 replied:
See earlier comments. PR #11 is the first of this splitting.. It does not include the glossary and most other files.

ianbush commented on docs/srcfixes.md:
Can we rename this to `common_issues.md` please.

mc7 replied:
This will be in a different PR. Leaving unresolved as reminder of file name change.

mc7 replied:
Have made note elsewhere. Resolving.

ianbush reviewed (changes requested):
Things to check:
- [x] Remove the stuff in `reposhow`
- [x] Rename `emptyj` to `example_schema` and `emptyjref` to `example_reference_schema`
- [x] Remove the `readme.md` in shared
- [x] Move glossary and tutorials to a separate branch
- [x] Some doc fixes

mc7 commented on docs/readme.md:
```diff
@@ -0,0 +1,109 @@
+
+A shortcut for step 3 is command `npm run build-all` but more
+targeted commands exist that are recommended at certain
+times. The tutorials use those other commands.
```
Have repharased in PR11 because the build-all command can fail on new schema using $ref keyword. The rephase explains when and two solutions.

mc7 commented on docs/readme.md:
```diff
@@ -0,0 +1,109 @@
+   repository. If issues are shown it will often be
+   quicker to run this locally whilst editing file(s).
+
+5. Populate current file with the unique parts of a
```
I've rephrased this in PR11. (The number were changed to letters after this comment so the numbered steps in tutorial 1 can refer back.)

There is a good reason for separating 2 and 5. In particular, the same instructions can be used for new schema and modifying existing schema where all previous releases must never break. It also splits issues caused by not
understanding the jsonschema-tools requirements  from issues caused by the JSON schema.

mc7 commented on docs/readme.md:
```diff
@@ -0,0 +1,109 @@
+and security standard. For example, are $ref to a particular
+schema release still valid?
+
+## Tutorials
```
Have started splitting this PR into smaller branches from main. The first is PR #11 which has the files needed to make the tools work and most of the other stuff removed.

mc7 commented on docs/readme.md:
```diff
@@ -0,0 +1,109 @@
+targeted commands exist that are recommended at certain
+times. The tutorials use those other commands.
+
+Step 4 is the command is `npm run check` which confirms every
```
The SRC means something very exact in context of schema whilst "check" is generic that likely to crop up during PRs with relation to "title",  "$id" and "$ref" values  I've left SRC for now in branch used by PR #11  and will resolve this conversation. But, can revisit in the new PR if needed.

mc7 commented on docs/readme.md:
```diff
@@ -0,0 +1,109 @@
+
+There are six steps, in summary:
+
+1. Create a schema folder inside schema root (the "schemas" folder)
```
PR #11 has the gist of this change but word "folder" used and the example changed so the schema is being put under an instrument folder. It seems doubtful an instrument will only have a single schema.
